File: 22_Change/python/change.py
"""
CHANGE

Change calculator

Port by Dave LeCompte
"""

PAGE_WIDTH = 64


def print_centered(msg):
    spaces = " " * ((PAGE_WIDTH - len(msg)) // 2)
    print(spaces + msg)


def print_header(title):
    print_centered(title)
    print_centered("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print()
    print()
    print()


def print_introduction():
    print("I, YOUR FRIENDLY MICROCOMPUTER, WILL DETERMINE")
    print("THE CORRECT CHANGE FOR ITEMS COSTING UP TO $100.")
    print()
    print()


def pennies_to_dollar_string(p):
    d = p / 100
    ds = f"${d:0.2f}"
    return ds


def compute_change():
    print("COST OF ITEM?")
    cost = float(input())
    print("AMOUNT OF PAYMENT?")
    payment = float(input())

    change_in_pennies = round((payment - cost) * 100)
    if change_in_pennies == 0:
        print("CORRECT AMOUNT, THANK YOU.")
        return

    if change_in_pennies < 0:
        short = -change_in_pennies / 100

        print(f"SORRY, YOU HAVE SHORT-CHANGED ME ${short:0.2f}")
        print()
        return

    print(f"YOUR CHANGE, {pennies_to_dollar_string(change_in_pennies)}")

    d = change_in_pennies // 1000
    if d > 0:
        print(f"{d} TEN DOLLAR BILL(S)")
    change_in_pennies -= d * 1000

    e = change_in_pennies // 500
    if e > 0:
        print(f"{e} FIVE DOLLAR BILL(S)")
    change_in_pennies -= e * 500

    f = change_in_pennies // 100
    if f > 0:
        print(f"{f} ONE DOLLAR BILL(S)")
    change_in_pennies -= f * 100

    g = change_in_pennies // 50
    if g > 0:
        print("ONE HALF DOLLAR")
    change_in_pennies -= g * 50

    h = change_in_pennies // 25
    if h > 0:
        print(f"{h} QUARTER(S)")
    change_in_pennies -= h * 25

    i = change_in_pennies // 10
    if i > 0:
        print(f"{i} DIME(S)")
    change_in_pennies -= i * 10

    j = change_in_pennies // 5
    if j > 0:
        print(f"{j} NICKEL(S)")
    change_in_pennies -= j * 5

    if change_in_pennies > 0:
        print(f"{change_in_pennies} PENNY(S)")


def print_thanks():
    print("THANK YOU, COME AGAIN.")
    print()
    print()


def main():
    print_header("CHANGE")
    print_introduction()

    while True:
        compute_change()
        print_thanks()


if __name__ == "__main__":
    main()

File: 22_Change/java/src/Change.java
import java.util.Arrays;
import java.util.Scanner;

/**
 * Game of Change
 * <p>
 * Based on the Basic game of Change here
 * https://github.com/coding-horror/basic-computer-games/blob/main/22%20Change/change.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's Basic game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 */
public class Change {

    // Used for keyboard input
    private final Scanner kbScanner;

    private enum GAME_STATE {
        START_GAME,
        INPUT,
        CALCULATE,
        END_GAME,
        GAME_OVER
    }

    // Current game state
    private GAME_STATE gameState;

    // Amount of change needed to be given
    private double change;

    public Change() {
        kbScanner = new Scanner(System.in);

        gameState = GAME_STATE.START_GAME;
    }

    /**
     * Main game loop
     */
    public void play() {

        do {
            switch (gameState) {
                case START_GAME:
                    intro();
                    gameState = GAME_STATE.INPUT;
                    break;

                case INPUT:

                    double costOfItem = displayTextAndGetNumber("COST OF ITEM ");
                    double amountPaid = displayTextAndGetNumber("AMOUNT OF PAYMENT ");
                    change = amountPaid - costOfItem;
                    if (change == 0) {
                        // No change needed
                        System.out.println("CORRECT AMOUNT, THANK YOU.");
                        gameState = GAME_STATE.END_GAME;
                    } else if (change < 0) {
                        System.out.println("YOU HAVE SHORT-CHANGES ME $" + (costOfItem - amountPaid));
                        // Don't change game state so it will loop back and try again
                    } else {
                        // Change needed.
                        gameState = GAME_STATE.CALCULATE;
                    }
                    break;

                case CALCULATE:
                    System.out.println("YOUR CHANGE, $" + change);
                    calculateChange();
                    gameState = GAME_STATE.END_GAME;
                    break;

                case END_GAME:
                    System.out.println("THANK YOU, COME AGAIN");
                    System.out.println();
                    gameState = GAME_STATE.INPUT;
            }
        } while (gameState != GAME_STATE.GAME_OVER);
    }

    /**
     * Calculate and output the change required for the purchase based on
     * what money was paid.
     */
    private void calculateChange() {

        double originalChange = change;

        int tenDollarBills = (int) change / 10;
        if (tenDollarBills > 0) {
            System.out.println(tenDollarBills + " TEN DOLLAR BILL(S)");
        }
        change = originalChange - (tenDollarBills * 10);

        int fiveDollarBills = (int) change / 5;
        if (fiveDollarBills > 0) {
            System.out.println(fiveDollarBills + " FIVE DOLLAR BILL(S)");
        }
        change = originalChange - (tenDollarBills * 10 + fiveDollarBills * 5);

        int oneDollarBills = (int) change;
        if (oneDollarBills > 0) {
            System.out.println(oneDollarBills + " ONE DOLLAR BILL(S)");
        }
        change = originalChange - (tenDollarBills * 10 + fiveDollarBills * 5 + oneDollarBills);

        change = change * 100;
        double cents = change;

        int halfDollars = (int) change / 50;
        if (halfDollars > 0) {
            System.out.println(halfDollars + " ONE HALF DOLLAR(S)");
        }
        change = cents - (halfDollars * 50);

        int quarters = (int) change / 25;
        if (quarters > 0) {
            System.out.println(quarters + " QUARTER(S)");
        }

        change = cents - (halfDollars * 50 + quarters * 25);

        int dimes = (int) change / 10;
        if (dimes > 0) {
            System.out.println(dimes + " DIME(S)");
        }

        change = cents - (halfDollars * 50 + quarters * 25 + dimes * 10);

        int nickels = (int) change / 5;
        if (nickels > 0) {
            System.out.println(nickels + " NICKEL(S)");
        }

        change = cents - (halfDollars * 50 + quarters * 25 + dimes * 10 + nickels * 5);

        int pennies = (int) (change + .5);
        if (pennies > 0) {
            System.out.println(pennies + " PENNY(S)");
        }

    }

    private void intro() {
        System.out.println(simulateTabs(33) + "CHANGE");
        System.out.println(simulateTabs(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
        System.out.println();
        System.out.println("I, YOUR FRIENDLY MICROCOMPUTER, WILL DETERMINE");
        System.out.println("THE CORRECT CHANGE FOR ITEMS COSTING UP TO $100.");
        System.out.println();
    }

    /*
     * Print a message on the screen, then accept input from Keyboard.
     * Converts input to a Double
     *
     * @param text message to be displayed on screen.
     * @return what was typed by the player.
     */
    private double displayTextAndGetNumber(String text) {
        return Double.parseDouble(displayTextAndGetInput(text));
    }

    /*
     * Print a message on the screen, then accept input from Keyboard.
     *
     * @param text message to be displayed on screen.
     * @return what was typed by the player.
     */
    private String displayTextAndGetInput(String text) {
        System.out.print(text);
        return kbScanner.next();
    }

    /**
     * Simulate the old basic tab(xx) command which indented text by xx spaces.
     *
     * @param spaces number of spaces required
     * @return String with number of spaces
     */
    private String simulateTabs(int spaces) {
        char[] spacesTemp = new char[spaces];
        Arrays.fill(spacesTemp, ' ');
        return new String(spacesTemp);
    }
}
